From ea4f1f7d0091848fde4a71f269770148515bb756 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Wed, 6 Nov 2024 15:04:32 +0100
Subject: [PATCH 706/841] feat(stm32mp2): allow to disable debug

Handle TS GEN only when CK_SYSDBG is enabled.
TS_GEN is clocked by APB-D

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 995b0a393c6e86944f47e7c4c548f87d3dfa9b0f feat(stm32mp2): allow to disable debug]
Change-Id: Ib07bf7e95d2d12934acdbf73552444b44f4a84a9
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/416842
Domain-Review: Maxime MERE <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/423128
Tested-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_pm.c | 27 +++++++++++++++++++--------
 1 file changed, 19 insertions(+), 8 deletions(-)

diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 6d67fe7d7..1ea01e1c5 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -217,29 +217,40 @@ static void stm32mp_state_set(unsigned int core_id, unsigned int state_id, bool
 static void stm32mp_ca35_lpi_isolate(void)
 {
 	/* Use write clear registers to clear bits */
-	mmio_write_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WC1), TS_CSYSREQ);
 	mmio_write_32(A35SSC_BASE + CA35SS_SSC_LPI_STGEN_NTS(WC1), STGEN_CSYSREQ);
 
-	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WC1)) & TS_CSYSACK) != 0U) {
+	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_STGEN_NTS(WC1))
+		& STGEN_CSYSACK) != 0U) {
 		;
 	}
 
-	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_STGEN_NTS(WC1)) & STGEN_CSYSACK) != 0U) {
-		;
+	if (clk_is_enabled(CK_SYSDBG)) {
+		mmio_write_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WC1), TS_CSYSREQ);
+
+		while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WC1))
+			& TS_CSYSACK) != 0U) {
+			;
+		}
 	}
 }
 
 static void stm32mp_ca35_lpi_restore(void)
 {
 	/* Use write set registers to set bits */
-	mmio_write_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WS1), TS_CSYSREQ);
 	mmio_write_32(A35SSC_BASE + CA35SS_SSC_LPI_STGEN_NTS(WS1), STGEN_CSYSREQ);
 
-	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WS1)) & TS_CSYSACK) == 0U) {
+	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_STGEN_NTS(WS1))
+		& STGEN_CSYSACK) == 0U) {
 		;
 	}
-	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_STGEN_NTS(WS1)) & STGEN_CSYSACK) == 0U) {
-		;
+
+	if (clk_is_enabled(CK_SYSDBG)) {
+		mmio_write_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WS1), TS_CSYSREQ);
+
+		while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_LPI_TSGEN_NTS(WS1))
+			& TS_CSYSACK) == 0U) {
+			;
+		}
 	}
 }
 
-- 
2.47.2

